crypto/aes/asm/aes-riscv64-zvkned.pl

```diff
@@ -0,0 +1,1413 @@
+    @{[vaeskf1_vi $v17, $v16, 7]}   # v17 <- rk8  (w[28,31])
+    @{[vaeskf1_vi $v18, $v17, 8]}   # v18 <- rk9  (w[32,35])
+    @{[vaeskf1_vi $v19, $v18, 9]}   # v19 <- rk10 (w[36,39])
+    @{[vaeskf1_vi $v20, $v19, 10]}  # v20 <- rk11 (w[40,43])
+
+    # Store the round keys
+    @{[vse32_v $v10, ($KEYP)]}
+    addi $KEYP, $KEYP, 16
+    @{[vse32_v $v11, ($KEYP)]}
+    addi $KEYP, $KEYP, 16
+    @{[vse32_v $v12, ($KEYP)]}
+    addi $KEYP, $KEYP, 16
+    @{[vse32_v $v13, ($KEYP)]}
+    addi $KEYP, $KEYP, 16
+    @{[vse32_v $v14, ($KEYP)]}
+    addi $KEYP, $KEYP, 16
+    @{[vse32_v $v15, ($KEYP)]}
+    addi $KEYP, $KEYP, 16
+    @{[vse32_v $v16, ($KEYP)]}
+    addi $KEYP, $KEYP, 16
+    @{[vse32_v $v17, ($KEYP)]}
+    addi $KEYP, $KEYP, 16
+    @{[vse32_v $v18, ($KEYP)]}
+    addi $KEYP, $KEYP, 16
+    @{[vse32_v $v19, ($KEYP)]}
+    addi $KEYP, $KEYP, 16
+    @{[vse32_v $v20, ($KEYP)]}
+
+    li a0, 1
+    ret
+.size L_set_key_128,.-L_set_key_128
+___
+
+$code .= <<___;
+.p2align 3
+L_set_key_256:
+    # Store the number of rounds
+    li $T1, 14
+    sw $T1, 240($KEYP)
+
+    @{[vsetivli__x0_4_e32_m1_tu_mu]}
+
+    # Load the key
+    @{[vle32_v $v10, ($UKEY)]}
+    addi $UKEY, $UKEY, 16
+    @{[vle32_v $v11, ($UKEY)]}
+
+    @{[vmv_v_v $v12, $v10]}
+    @{[vaeskf2_vi $v12, $v11, 2]}
+    @{[vmv_v_v $v13, $v11]}
+    @{[vaeskf2_vi $v13, $v12, 3]}
+    @{[vmv_v_v $v14, $v12]}
+    @{[vaeskf2_vi $v14, $v13, 4]}
+    @{[vmv_v_v $v15, $v13]}
+    @{[vaeskf2_vi $v15, $v14, 5]}
+    @{[vmv_v_v $v16, $v14]}
+    @{[vaeskf2_vi $v16, $v15, 6]}
+    @{[vmv_v_v $v17, $v15]}
+    @{[vaeskf2_vi $v17, $v16, 7]}
+    @{[vmv_v_v $v18, $v16]}
+    @{[vaeskf2_vi $v18, $v17, 8]}
+    @{[vmv_v_v $v19, $v17]}
+    @{[vaeskf2_vi $v19, $v18, 9]}
+    @{[vmv_v_v $v20, $v18]}
+    @{[vaeskf2_vi $v20, $v19, 10]}
+    @{[vmv_v_v $v21, $v19]}
+    @{[vaeskf2_vi $v21, $v20, 11]}
+    @{[vmv_v_v $v22, $v20]}
+    @{[vaeskf2_vi $v22, $v21, 12]}
+    @{[vmv_v_v $v23, $v21]}
+    @{[vaeskf2_vi $v23, $v22, 13]}
+    @{[vmv_v_v $v24, $v22]}
+    @{[vaeskf2_vi $v24, $v23, 14]}
+
+    @{[vse32_v $v10, ($KEYP)]}
+    addi $KEYP, $KEYP, 16
+    @{[vse32_v $v11, ($KEYP)]}
+    addi $KEYP, $KEYP, 16
+    @{[vse32_v $v12, ($KEYP)]}
+    addi $KEYP, $KEYP, 16
+    @{[vse32_v $v13, ($KEYP)]}
+    addi $KEYP, $KEYP, 16
+    @{[vse32_v $v14, ($KEYP)]}
+    addi $KEYP, $KEYP, 16
+    @{[vse32_v $v15, ($KEYP)]}
+    addi $KEYP, $KEYP, 16
+    @{[vse32_v $v16, ($KEYP)]}
+    addi $KEYP, $KEYP, 16
+    @{[vse32_v $v17, ($KEYP)]}
+    addi $KEYP, $KEYP, 16
+    @{[vse32_v $v18, ($KEYP)]}
+    addi $KEYP, $KEYP, 16
+    @{[vse32_v $v19, ($KEYP)]}
+    addi $KEYP, $KEYP, 16
+    @{[vse32_v $v20, ($KEYP)]}
+    addi $KEYP, $KEYP, 16
+    @{[vse32_v $v21, ($KEYP)]}
+    addi $KEYP, $KEYP, 16
+    @{[vse32_v $v22, ($KEYP)]}
+    addi $KEYP, $KEYP, 16
+    @{[vse32_v $v23, ($KEYP)]}
+    addi $KEYP, $KEYP, 16
+    @{[vse32_v $v24, ($KEYP)]}
```

Comment:
Why not interleaved the stores `vse32` and key schedule expand `vaeskf2.vi` ?

```
    @{[vmv_v_v $v12, $v10]}
    @{[vaeskf2_vi $v12, $v11, 2]}
    @{[vmv_v_v $v13, $v11]}
    @{[vaeskf2_vi $v13, $v12, 3]}
    @{[vmv_v_v $v14, $v12]}
    @{[vaeskf2_vi $v14, $v13, 4]}
    @{[vmv_v_v $v15, $v13]}
    (...)
    @{[vse32_v $v10, ($KEYP)]}
    addi $KEYP, $KEYP, 16
    @{[vse32_v $v11, ($KEYP)]}
    addi $KEYP, $KEYP, 16
```
would become
````
    # index 0 key round store
    @{[vse32_v $v10, ($KEYP)]}
    addi $KEYP, $KEYP, 16
    # index 1 key round store
    @{[vse32_v $v11, ($KEYP)]}
    addi $KEYP, $KEYP, 16

    # index 2 key round and store
    # before (0: v10, 1: v11)
    @{[vaeskf2_vi $v12, $v11, 2]}
    # after (0: v10, 1: v11, 2: v12)
    @{[vse32_v $v12, ($KEYP)]}
    addi $KEYP, $KEYP, 16

    # index 3 key round and store
    # before (0: v10, 1: v11, 2: v12)
    @{[vaeskf2_vi $v11, $v12, 2]}
    # after (0: v10, 2: v12, 3: v11)
    @{[vse32_v $v11, ($KEYP)]}
    addi $KEYP, $KEYP, 16

    # index 4 key round and store
    # before (0: v10, 2: v12, 3: v11)
    @{[vaeskf2_vi $v12, $v11, 2]}
    # after (0: v10, 3: v11, 4: v12)
    @{[vse32_v $v12, ($KEYP)]}
    addi $KEYP, $KEYP, 16

    (...)
```

This saves the `vmv` and it does not look like having the key in v10-v22 is required after this routine (but I may have missed something)

Reply:
I agree we can interleave it, but I think the round number for vaeskf2_vi should be increased and "$v12" should be changed to "$v10". Overall, this is a very good suggestion.

Reply:
Our benchmarks showed no performance improvement with the new method, as the new method seems to have more dependencies between instructions. As the result, the implementation remains the same.